### PR TITLE
kinetis: Update i2c driver to new API

### PR DESCRIPTION
**Pull request by gebart**
### Contribution description

Update the Kinetis I2C driver for the new API

Needs some more testing

### Issues/PRs references

#6577

### CODE REVIEW

File: cpu/kinetis/periph/i2c.c [hunk 1]
```diff
@@ -45,474 +48,322 @@
...
+      28,   32,   36,   40,   44,   48,   56,   68, /* 0x08 ~ 0x0f */
+      48,   56,   64,   72,   80,   88,  104,  128, /* 0x10 ~ 0x17 */
+      80,   96,  112,  128,  144,  160,  192,  240, /* 0x18 ~ 0x1f */
+     160,  192,  224,  256,  288,  320,  384,  480, /* 0x20 ~ 0x27 */
+     320,  384,  448,  512,  576,  640,  768,  960, /* 0x28 ~ 0x2f */
+     640,  768,  896, 1024, 1152, 1280, 1536, 1920, /* 0x30 ~ 0x37 */
+    1280, 1536, 1792, 2048, 2304, 2560, 3072, 3840, /* 0x38 ~ 0x3f */
 };
 
+static const i2c_conf_t i2c_config[I2C_NUMOF] = I2C_CONFIG;
```
Comment (aabadie) [inline cpu/kinetis/periph/i2c.c]: Why not directly defining this array of structs in the board periph_conf (like stm32, sam0, etc) instead of using the I2C_CONFIG macro ?

Comment (gebart) [reply cpu/kinetis/periph/i2c.c]: I wanted to avoid unintentionally instancing this local array inside any other files, so I decided to put it as a define instead.

Comment (aabadie) [reply cpu/kinetis/periph/i2c.c]: Ok but then it's not consistent with the other configurations arrays (UART, SPI, PWN, ADC, etc).

**Comment by gebart** [comment]
Cherry picked in #8933 #8814 #8930 to avoid rebase problems later

File: cpu/kinetis/periph/i2c.c (hunk 1 above)
Comment (gebart) [reply cpu/kinetis/periph/i2c.c]: alright, will change this for now

**Comment by gebart** [comment]
Refactored to use IRQs when transferring the actual data bytes. This method lets the CPU perform other tasks while waiting for a slow I2C slave.

File: boards/teensy31/Makefile.features [hunk 2]
```diff
@@ -1,11 +1,13 @@
 # Put defined MCU peripherals here (in alphabetical order)
-FEATURES_PROVIDED += periph_cpuid
-FEATURES_PROVIDED += periph_gpio
 FEATURES_PROVIDED += periph_pwm
+FEATURES_PROVIDED += periph_rtc
+FEATURES_PROVIDED += periph_rtt
```
Comment (dylad) [inline boards/teensy31/Makefile.features]: Is this related ?

File: cpu/kinetis/periph/i2c.c [hunk 3]
```diff
@@ -45,474 +51,502 @@
...
-    int n = 0;
-
-    while (length > 0) {
-        TRACE("i2c: tx: %02x\n", data[n]);
-        dev->D = data[n];
-
-        while (!(dev->S & I2C_S_IICIF_MASK));
-
-        dev->S = I2C_S_IICIF_MASK;
+    I2C_Type *i2c = i2c_config[dev].i2c;
```
Comment (dylad) [inline cpu/kinetis/periph/i2c.c]: Could you add a guard here please ? 
```
if (len == 0 || data == NULL) {
    return -EINVAL;
}
```

Comment (dylad) [reply cpu/kinetis/periph/i2c.c]: Could you also return in case of unsupported features ?

File: boards/teensy31/Makefile.features (hunk 2 above)
Comment (gebart) [reply boards/teensy31/Makefile.features]: I had to pull in some upstream merged PRs to avoid problems rebasing later.

File: cpu/kinetis/periph/i2c.c (hunk 3 above)
Comment (gebart) [reply cpu/kinetis/periph/i2c.c]: `len == 0` is not an error, it can be used to send a start condition or stop condition, depending on flags.
I don't think there are any unsupported flags here (other than REG16, which I think we can just ignore in the _bytes functions)

Comment (dylad) [reply cpu/kinetis/periph/i2c.c]: > it can be used to send a start condition or stop condition, depending on flags

I never tought of it but is this really useful for the user ? I mean, we're suppose to read or write on the bus so why just sending a start out of the blue ?

Comment (gebart) [reply cpu/kinetis/periph/i2c.c]: The hih6130 is one weird device which uses empty writes to trigger measurements. The user sends a start condition for writing followed by a stop condition, so zero data bytes. Then when reading the device will return the most recent measurement.

Comment (dylad) [reply cpu/kinetis/periph/i2c.c]: > The hih6130 is one weird device which uses empty writes to trigger measurements

Indeed, I'll look at this one. Is this even I2C/TWI compliant ?

Comment (gebart) [reply cpu/kinetis/periph/i2c.c]: Dunno, but it's a device that we have a driver for

**Comment by gebart** [comment]
It would be a good idea to test with the hih6130 as well because of the null write behaviour, but I don't remember where I put my breakout board for that sensor..

**Comment by gebart** [comment]
Added some asserts for enabling interrupts and a check to ensure that data != NULL when len > 0

**Comment by dylad** [comment]
@gebart I tend to ACK this PR but I have no hardware to test it :(
Do you know if someone have the required hardware for testing ?

**Comment by gebart** [comment]
I have Mulle and frdm boards, and I know that there are some kinetis boards among the FU Berlin people

**Comment by dylad** [comment]
I guess I'll resend an email on the devel mailing to require some help for testing

**Comment by MichelRottleuthner** [comment]
Just tested this with pba-d-01-kw2x and the already merged srf08 driver. Works fine

File: cpu/kinetis/periph/i2c.c [hunk 4]
```diff
@@ -45,474 +51,515 @@
...
-#if I2C_1_EN
-    [I2C_1] = MUTEX_INIT,
-#endif
-#if I2C_2_EN
-    [I2C_2] = MUTEX_INIT
-#endif
-#if I2C_3_EN
-    [I2C_3] = MUTEX_INIT
-#endif
+#define THREAD_FLAG_KINETIS_I2C (1u << 8)
```
Comment (gebart) [inline cpu/kinetis/periph/i2c.c]: Added this to avoid dependency on contested PR #9284. We can update the driver later if a consensus is reached regarding the reserved thread flags

**Comment by dylad** [comment]
Just re-trigger Travis and we're good to go.

File: cpu/kinetis/periph/i2c.c [hunk 5]
```diff
@@ -45,474 +51,515 @@
...
+        i2c_state[dev].retval = 0;
+        bit_set8(&i2c->C1, I2C_C1_IICIE_SHIFT);
+        /* Initiate master receive mode by reading the data register once when
+         * the C1[TX] bit is cleared and C1[MST] is set */
+        volatile uint8_t dummy;
+        dummy = i2c->D;
+        ++dummy;
+        /* Wait until the ISR signals back */
+        TRACE("i2c: read C1=%02x S=%02x\n", (unsigned)i2c->C1, (unsigned)i2c->S);
+        thread_flags_t tflg = thread_flags_wait_any(THREAD_FLAG_KINETIS_I2C | THREAD_FLAG_TIMEOUT);
```
Comment (gebart) [inline cpu/kinetis/periph/i2c.c]: We are checking for the TIMEOUT thread flag, but there is not yet anything setting it in the driver. I left this thread flag check here to make it simple in the future to add a timeout, but I did not yet find a satisfying solution which does not add extra dependencies to the driver. This part of the code can still be used if a user wants to eliminate the possibility of a bus hang blocking the application, they can use xtimer to provide a timer which sets the timeout flag.

Comment (dylad) [reply cpu/kinetis/periph/i2c.c]: I'm fine with this, switching thread while waiting here is an excellent idea and IMHO we should do it for every MCU/drivers :)

**Comment by dylad** [comment]
Here we go ! thanks @gebart for your work !
